api/resolvers/price.js

```diff
@@ -0,0 +1,32 @@
+const cache = new Map();
+const expiresIn = 30000; // in milliseconds
+
+async function getPrice(fiat) {
+  fiat ??= 'USD';
+  if (cache.has(fiat)) {
+    const { price, createdAt } = cache.get(fiat)
+    const expired = createdAt + expiresIn < Date.now()
+    if (!expired) {
+      return price;
+    }
+  }
+  const url = `https://api.coinbase.com/v2/prices/BTC-${fiat}/spot`;
```

Comment:
Without a lock here, in the worst case, every visitor would cause a coinbase API call if every request notices that the cache is expired at the same time and then goes on to do a API call.

@huumn, do you think a lock is thus required here?

Reply:
Thinking deeper about this and if this is indeed possible since NodeJS event loop is single-threaded:

I still think it's still possible that this happens due to how async/await works.
While we are fetching (which uses `await fetch`), the scheduler continues with another event which could invoke `getPrice` and see that the cache entry is (still) expired. It will then go to `await fetch` and give up execution to the scheduler. Therefore, two API calls will happen. This could happen an "infinite amount of times"

Reply:
Worst case here seems to be that we make so many calls to coinbase at the same time, we get rate limited and responses from coinbase fail temporarily. I'd guess the probability of that happening is near zero, and so long as we can gracefully handle coinbase failing (in the rare event that it would occur), there's no need for a lock.